### PR TITLE
Add Docker build jobs

### PR DESCRIPTION
**Pull request by alunsin**
Add 4 jobs for building Docker and containerd:
- postsubmit-build-docker, build docker packages and trigger postsubmit-build-containerd 
- postsubmit-build-containerd, build containerd and run the validation tests
- postsubmit-test-docker-staging, test the packages from the docker's staging repo
- postsubmit-test-docker-release, test the packages from the docker's release repo

### CODE REVIEW

**Comment by alunsin** [comment]
Please find the job results from the test node:

- https://prow.ppc64le-cloud.org/view/s3/prow-logs/logs/postsubmit-build-docker
- https://prow.ppc64le-cloud.org/view/s3/prow-logs/logs/postsubmit-build-containerd
- https://prow.ppc64le-cloud.org/view/s3/prow-logs/logs/postsubmit-test-docker-staging
- https://prow.ppc64le-cloud.org/view/s3/prow-logs/logs/postsubmit-test-docker-release

**Comment by alunsin** [comment]
Review comment updates:
- Only report slack notification on failure for postsubmit-build-dockr
- Renamed postsubmit-build-containerd to postsubmit-build-containerd
- Renamed prow-test-docker.sh to prow-build-test-containerd.sh
- Moved package repository based test jobs to postsubmit-test-repo-docker.yaml files

Also tested the jobs, see details:
 *  https://prow.ppc64le-cloud.org/view/s3/prow-logs/logs/postsubmit-build-docker
 * https://prow.ppc64le-cloud.org/view/s3/prow-logs/logs/postsubmit-build-test-containerd
 * https://prow.ppc64le-cloud.org/view/s3/prow-logs/logs/postsubmit-test-docker-staging
 * https://prow.ppc64le-cloud.org/view/s3/prow-logs/logs/postsubmit-test-docker-release

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+          args:
+          - -c
+          - |
+            set -o errexit
+            set -o nounset
+            set -o pipefail
+            set -o xtrace
+
+            #Timestamp file from the tracking branch (required to access the proper s3 directory)
+            TIMESTAMP_FILE=postsubmit-build-docker
```
Comment (Rajalakshmi-Girish) [inline config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml:38]: I think these jobs are for testing the binaries from the Docker community's stagging and release websites.
But the comment here says 
`required to access the proper s3 directory` can you please clarify why exactly Timestamp file is needed.
Also the need for fetching `S3_SECRET_AUTH` from `docker-s3-credentials` below?

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+           secretName: git-ssh-docker-ce-build
+
+    - name: postsubmit-build-test-containerd
+      cluster: k8s-ppc64le-cluster
+      decorate: true
+      decoration_config:
+        timeout: 3h
+      run_if_changed: 'job/postsubmit-build-docker'
+      branches:
+        - prow-job-tracking
```
Comment (Rajalakshmi-Girish) [inline config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml:76]: @alunsin Can you please explain the need to have a separate branch for prow job tracking?

IMO, we could reduce the length of these prow job yamls and make it simple if the `TIMESTAMP_FILE` is in the main branch itself.

**Review by Rajalakshmi-Girish — CHANGES_REQUESTED** [review]

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+      cluster: k8s-ppc64le-cluster
+      decorate: true
+      decoration_config:
+        timeout: 3h
+      branches:
+          - main
+      run_if_changed: 'env/env.list'
+      reporter_config:
+        slack:
+          channel: 'prow-job-notifications'
```
[the remark this answers is not in the document]
Comment (mkumatag) [reply config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml:13]: I think we can just merge and see how it goes, worst case will throw some messages :)

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+            echo "* Setup the files from git *"
+
+            echo "* Backup timestamp file from the previous job *"
+            cp job/${TIMESTAMP_FILE} /tmp
+
+            echo "* Switch back to the git branch: ${REPO_OWNER}/${REPO_NAME}/${WORK_BRANCH} *"
+            git fetch https://github.com/${REPO_OWNER}/${REPO_NAME} ${WORK_BRANCH}
+            git branch --force ${WORK_BRANCH} FETCH_HEAD
+            git checkout ${WORK_BRANCH}
```
[the remark this answers is not in the document]
Comment (alunsin) [reply config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml:124]: Unfortunately no, see point 4 in my [previous comment ](https://github.com/ppc64le-cloud/test-infra/pull/293/files/283b0b32c0d8dac9c388c2a6ef45b56906ac6246#r834270804) above

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml (hunk 1 above)
Comment (alunsin) [reply config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml:38]: Even if the packages are installed from the docker's websites, we still need to access date from our internal s3 directory,  In particular we need to know what distribution were built and do the test for all of them

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Rajalakshmi-Girish) [reply config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml:124]: @alunsin yes as you explained in the previous call, I understand that the job initially has the content populated from `prow-job-tracking` branch and you are switching to the main branch to access the scripts in it.

My question was,
Can we not skip L122 and L123 and directly say `git checkout ${WORK_BRANCH}` for switching to main branch?

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml (hunk 2 above)
Comment (Rajalakshmi-Girish) [reply config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml:76]: @alunsin I see the `date.list` file in the master branch at https://github.com/ppc64le-cloud/docker-ce-build/blob/main/env/date.list
created for a similar purpose as the `postsubmit-build-docker` file in the `prow-job-tracking` branch.
Would it not be needed to update `date.list` file in the `master` branch? (I see it having a old value)

In fact, the script at https://github.com/ppc64le-cloud/docker-ce-build/blob/main/prow-build-docker.sh#L58 is using the `date.list` file in the `main` branch as a temporary location before copying to the `prow-job-tracking` branch.

Also, the folder `env` in the `main` branch has only the job triggering files that we update to spin another prow job.

I see isolation already present in terms of having a separate folder. Would it be right to rename `env` folder in the main branch to `job`?

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml (hunk 1 above)
Comment (Rajalakshmi-Girish) [reply config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml:38]: IMO, files `env/test-staging.list` and `env/test-release.list` should be having the details of date, and distribution details that are to be tested from docker's websites rather than going back to your s3 directory.

The `env/test-release.list` can have a fixed list of distributions that need to be tested and can have a single DATE parameter to be changed manually whenever you want to run tests for packages from the website.

If the distribution list changes in the future with a date, you may change that too when you manually update DATE.

**Review by Rajalakshmi-Girish — CHANGES_REQUESTED** [review]

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (alunsin) [reply config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml:124]: No, that would not work. doing a git checkout <branch> will fail here, as the git branch is not fetched.
That's because Prow's clonerefs [pod utilities ](https://github.com/kubernetes/test-infra/tree/master/prow/cmd#pod-utilities) does a light/partial git clone as explained below and only the current branch (prow-job-tracking)  has been fetched by git.
Here is how `clonerefs` init container does the git clone ( as per the clonerefs container's log):
```
git init
git config user.name ci-robot
git config user.email ci-robot@k8s.io
git fetch https://github.com/alunsin/docker-ce-build.git  --tags --prune
git fetch https://github.com/alunsin/docker-ce-build.git 5ed27dcf83b3bc1425ec529436234a59f7e93fb7
git branch --force prow-job-tracking 5ed27dcf83b3bc1425ec529436234a59f7e93fb7
git checkout 5ed27dcf83b3bc1425ec529436234a59f7e93fb7
```

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml (hunk 1 above)
Comment (alunsin) [reply config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml:38]: I am not sure if I would agree to set the list of distributions manually in the .list files. 
FYI we are populating the list of distribution automatically by looking at the directory names contained in the  [docker-ce-packaging](https://github.com/docker/docker-ce-packaging/tree/master/deb) repository maintained by Docker. 
This list of distro is evolving quite a bit, and this part the automation has been working quite smoothly so far. So I would better keep this in place. 
I have reviewed the code and I think we calculate that list of distro all the times and do not have any dependency on s3 data for that purpose. So my bad on this.
However we are storing our logs somewhere as a proof of test on the s3 directory and I would like to keep this as it is if possible.

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml (hunk 2 above)
Comment (alunsin) [reply config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml:76]: The `date.list` is a bit messy in term of design and this something that we could review and improve in the future.
I need to double check, but I am  pretty sure the `env/date.list` file can be removed from git.

If you look at [https://github.com/ppc64le-cloud/docker-ce-build/blob/87fc960fa9acffb91e8e55f9c7529582618479e5/prow-build-docker.sh#L10](/prow-build-docker.sh) , the date.list file is updating locally  in the `postsubmit-build-docker` job, which is the 1st job that will be spawned.
Then we transfer this date content from date.list file to the secondaries jobs by copying this content into the file postsubmit-build-docker in the `prow-job-tracking` branch. The other jobs uses that file to populate the date.list file and make sure they are all using the same s3 directory.
To sum up I would prefer removing the `env/date.list` from the git repo. Will test that as soon as I am done with the current docker refresh.

**Review by Rajalakshmi-Girish — COMMENTED** [review]
/lgtm @mkumatag Can you please take a look.

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+            echo "* Restore back to the time stamp file from the tracking branch *"
+            cp /tmp/${TIMESTAMP_FILE} env/date.list
+            cat env/date.list
```
Comment (mkumatag) [inline config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml:128]: why are we cating the file here? what's the content of this file?

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+      cluster: k8s-ppc64le-cluster
+      decorate: true
+      run_if_changed: 'env/test-staging.list'
+      branches:
+        - main
+      reporter_config:
+        slack:
+          channel: 'prow-job-notifications'
+          job_states_to_report:
+           - success
```
Comment (mkumatag) [inline config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml:13]: same comment as above

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+      cluster: k8s-ppc64le-cluster
+      decorate: true
+      run_if_changed: 'env/test-release.list'
+      branches:
+        - main
+      reporter_config:
+        slack:
+          channel: 'prow-job-notifications'
+          job_states_to_report:
+           - success
```
Comment (mkumatag) [inline config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml:77]: same comment as above

**Review by mkumatag — CHANGES_REQUESTED** [review]
few minor comments

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml (hunk 7 above)
Comment (Rajalakshmi-Girish) [reply config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml:77]: @alunsin mentioned these jobs would be manually triggered by chnging files [test-release.list](https://github.com/ppc64le-cloud/docker-ce-build/blob/main/env/test-release.list) and [test-staging.list](https://github.com/ppc64le-cloud/docker-ce-build/blob/main/env/test-staging.list) when the ppc builds are uploaded to docker website by community people.

Hope there would be no flooding of success messages in channel.

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+      decoration_config:
+        timeout: 3h
+      run_if_changed: 'job/postsubmit-build-docker'
+      branches:
+        - prow-job-tracking
+      reporter_config:
+        slack:
+          channel: 'prow-job-notifications'
+          job_states_to_report:
+           - success
```
[the remark this answers is not in the document]
Comment (mkumatag) [reply config/jobs/ppc64le-cloud/build-docker/postsubmit-build-docker.yaml:81]: alright, with this assumption - we can merge the code.

**Review by mkumatag — APPROVED** [review]
/lgtm
/approve

File: config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml (hunk 7 above)
Comment (alunsin) [reply config/jobs/ppc64le-cloud/build-docker/postsubmit-test-repo-docker.yaml:77]: Sorry for the late reply. I am not expecting those jobs to flood the slack channel as those are manually triggered by editing the env/*.list files whenever we have a new docker and containerd version available from the upstream as already explained by @Rajalakshmi-Girish. Thanks all of your review!
